aioxmpp/ibr/service.py

```diff
@@ -0,0 +1,253 @@
+            )
+            reply = yield from aioxmpp.send_and_wait_for(
+                xmlstream, [iq], [aioxmpp.IQ]
+            )
+        return reply
+
+    @asyncio.coroutine
+    def get_registration_fields(self,
+                                peer_jid,
+                                timeout=60,
+                                no_verify=True
+                                ):
+        """
+        A query is sent to the server to obtain the fields that need to be
+        filled to register with the server.
+
+        :param peer_jid: Server where he query will be sent.
+        :type peer_jid: :class:`aioxmpp.JID`
+
+        :param timeout: Maximum time in seconds to wait for an IQ response, or
+                        :data:`None` to disable the timeout.
+        :type timeout: :class:`~numbers.Real` or :data:`None`
+
+        :param no_verify: Specifies whether the server's TLS is verified.
+        :type no_verify: :class:`aioxmpp.JID`
+
+        :return: :attr:`list`
+        """
+        iq = aioxmpp.IQ(
+            to=peer_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.GET,
+            payload=xso.Query()
+        )
+        iq.autoset_id()
+
+        loop = asyncio.get_event_loop()
+        metadata = aioxmpp.make_security_layer('', no_verify=no_verify)
+
+        try:
+            reply = yield from self.connect_and_send(iq, loop, metadata, peer_jid, timeout)
+
+            return [a for a in dir(reply.payload)
+                    if isinstance(getattr(reply.payload, a), str) and
+                    not a.startswith('__')]
+        except Exception as exc:
+            raise exc
+
+    @asyncio.coroutine
+    def register(self,
+                 jid,
+                 password,
+                 aux_fields=None,
+                 timeout=60,
+                 no_verify=True
+                 ):
+        """
+        Create a new account on the server.
+
+        :param jid: Server where he account will be created wih the username
+                    of the account.
+        :type jid: :class:`aioxmpp.JID`
+
+        :param password: Password of the new account.
+        :type password: :class:`str`
+
+        :param aux_fields: Auxiliary fields in case additional info is needed.
+        :type aux_fields: :class:`dict`
+
+        :param timeout: Maximum time in seconds to wait for an IQ response, or
+                        :data:`None` to disable the timeout.
+        :type timeout: :class:`~numbers.Real` or :data:`None`
+
+        :param no_verify: Specifies whether the server's TLS is verified.
+        :type no_verify: :class:`aioxmpp.JID`
+
+        :return: :attr:`list`
+        """
+        iq = aioxmpp.IQ(
+            to=jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.SET,
+            payload=xso.Query()
+        )
+        iq.autoset_id()
+
+        iq.payload.username = jid.localpart
+        iq.payload.password = password
+
+        if aux_fields is not None:
```

Comment:
Perhaps expose the XSO query class instead (for example by having an advanced method that takes an XSO instead of `jid` and `password`). Exposing XSOs across module boundaries is not per se considered bad style in aioxmpp. I don't think this solution is particularly bad, but I like to limit the use of `setattr` and the likes to the absolutely necessary.

Reply:
This also plays nice with the change suggested to ``get_registration_fields``.